### DISCUSSION TITLE
Why make incompatible changes to draw_networkx_edges

### DISCUSSION THREAD

**Discussion by jolespin** [networkx/networkx#5467] [Q&A] [answered]
Why drop linestyle for style and edges for edgelist? Is this permanent?

**Comment by rossbar** [comment]
Can you provide more information? For example, a brief code snippet that worked the way you expected in nx version 2.A.B that no longer works in 2.X.Y (along with the corresponding values for A, B, X, and Y)

**Comment by jolespin** [comment]
Sorry, that was a bit of a lazy question I asked earlier.


```
>>> import matplotlib.pyplot as plt
>>> import networkx as nx
>>> graph = nx.complete_graph(5)
>>> pos = nx.spring_layout(graph)
>>> nx.draw_networkx_edges(graph, pos=pos, edges=[0,1,2])
/Users/jespinoz/anaconda3/envs/test_networkx/lib/python3.7/site-packages/networkx/drawing/nx_pylab.py:563: MatplotlibDeprecationWarning:
The iterable function was deprecated in Matplotlib 3.1 and will be removed in 3.3. Use np.iterable instead.
  if not cb.iterable(width):
/Users/jespinoz/anaconda3/envs/test_networkx/lib/python3.7/site-packages/networkx/drawing/nx_pylab.py:611: MatplotlibDeprecationWarning:
The is_numlike function was deprecated in Matplotlib 3.0 and will be removed in 3.2. Use isinstance(..., numbers.Number) instead.
  if cb.is_numlike(alpha):
<matplotlib.collections.LineCollection object at 0x7fa44b3e3470>
>>> plt.show()
```

These are mostly old versions: 

```
# packages in environment at /Users/jespinoz/anaconda3/envs/test_networkx:
#
# Name                    Version                   Build  Channel
bzip2                     1.0.8                hc929b4f_4    conda-forge
ca-certificates           2022.3.29            hecd8cb5_0
cycler                    0.11.0             pyhd8ed1ab_0    conda-forge
decorator                 5.1.1              pyhd8ed1ab_0    conda-forge
freetype                  2.11.0               hd8bbffd_0
kiwisolver                1.4.2            py37h18621fa_1    conda-forge
libblas                   3.9.0           13_osx64_openblas    conda-forge
libcblas                  3.9.0           13_osx64_openblas    conda-forge
libcxx                    13.0.1               hc203e6f_0    conda-forge
libedit                   3.1.20210910         hca72f7f_0
libffi                    3.2.1                         1    bioconda
libgfortran               5.0.0           9_3_0_h6c81a4c_23    conda-forge
libgfortran5              9.3.0               h6c81a4c_23    conda-forge
liblapack                 3.9.0           13_osx64_openblas    conda-forge
libopenblas               0.3.18          openmp_h3351f45_0    conda-forge
libpng                    1.6.37               hb0a8c7a_2    conda-forge
libzlib                   1.2.11            h6c3fc93_1014    conda-forge
llvm-openmp               13.0.1               hcb1a161_1    conda-forge
matplotlib                3.1.0                    py37_1    conda-forge
matplotlib-base           3.1.0            py37h3a684a6_1    conda-forge
ncurses                   6.3                  he49afe7_0    conda-forge
networkx                  2.2                        py_1    conda-forge
numpy                     1.21.5           py37h3c8089f_0    conda-forge
openssl                   1.0.2u               h0b31af3_0    conda-forge
pip                       22.0.4             pyhd8ed1ab_0    conda-forge
pyparsing                 3.0.7              pyhd8ed1ab_0    conda-forge
python                    3.7.0             h145921a_1006    conda-forge
python-dateutil           2.8.2              pyhd8ed1ab_0    conda-forge
python_abi                3.7                     2_cp37m    conda-forge
readline                  7.0               hcfe32e1_1001    conda-forge
setuptools                62.0.0           py37hf985489_0    conda-forge
six                       1.16.0             pyh6c4a22f_0    conda-forge
sqlite                    3.33.0               hffcf06c_0
tk                        8.6.12               h5dbffcc_0    conda-forge
tornado                   6.1              py37h69ee0a8_3    conda-forge
typing-extensions         4.1.1                hd8ed1ab_0    conda-forge
typing_extensions         4.1.1              pyha770c72_0    conda-forge
wheel                     0.37.1             pyhd8ed1ab_0    conda-forge
xz                        5.2.5                haf1e3a3_1    conda-forge
zlib                      1.2.11            h6c3fc93_1014    conda-forge
```


Here's what happens why I try to run the same code using the newer NetworkX versions (v2.6.3):

```
>>> import matplotlib.pyplot as plt
>>> import networkx as nx
>>> graph = nx.complete_graph(5)
>>> pos = nx.spring_layout(graph)
>>> nx.draw_networkx_edges(graph, pos=pos, edges=[0,1,2])

# ---------------------------------------------------------------------------
# TypeError                                 Traceback (most recent call last)
# Input In [5], in <module>
#       3 graph = nx.complete_graph(5)
#       4 pos = nx.spring_layout(graph)
# ----> 5 nx.draw_networkx_edges(graph, pos=pos, edges=[0,1,2])

# TypeError: draw_networkx_edges() got an unexpected keyword argument 'edges'
```


The reason why I brought this up is because I have some calls to my packages that use draw_networkx_edges so now I need to change the code so they can work with newer versions. Was just asking why you decided to make a breaking change in a non-major release (e.g., 1.x vs. 2.x)? 

Kind regards,
-Josh

**Comment by rossbar** [comment, marked as answer]
Thanks for clarifying!

> Is this permanent?

tl;dr - yes, there is not plan to revert to the previous keyword argument names.

> The reason why I brought this up is because I have some calls to my packages that use draw_networkx_edges so now I need to change the code so they can work with newer versions. Was just asking why you decided to make a breaking change in a non-major release (e.g., 1.x vs. 2.x)?

Many of the scientific Python libraries (NetworkX included) don't strictly follow semantic versioning in the sense that breaking changes can only occur at major releases[^1]. Instead, NetworkX [has a (rough) policy](https://networkx.org/documentation/latest/developer/deprecations.html#policy) that deprecation warnings should be raised for ~2 minor releases (or 1 if the next release will be a major release) prior to any breaking change. These values roughly mirror Python's deprecation policy (warn for 2 minor releases before breaking changes). The ultimate goal is to strike a balance between improving the library and breaking existing code - absent user feedback (typically in the form of Github issues) developers don't really have any good way to gauge how well we're hitting that balance point for most users, so opening discussions like this is useful!

[^1]: There are a number of reasons for this --- one major reason is that most projects are largely maintained by volunteers in their free time and as a result, there may be long stretches of time where there is not that much activity on a project.

**Reply by jolespin** [reply]
Thank you for the help and the explanation. I'll change my code so it reflects the new change.

Is there a change log so I can see if any of my other code is affected? Also, what version did this change occur so I can make sure to update my deps

**Reply by rossbar** [reply]
> Is there a change log so I can see if any of my other code is affected?

The [Release notes](https://networkx.org/documentation/latest/release/index.html#releases) are the canonical place where the developers try to summarize all of the changes at each release.

> Also, what version did this change occur so I can make sure to update my deps

Looking more closely at your specific example, I don't think `edges=` has ever been a valid keyword argument for `draw_networkx_edges`, at least not AFAICT from a look at the `git blame`. Even if it were, `edges=[0, 1, 2]` would not be a valid edgelist, so it was likely always a bug that this didn't raise an error back in v2.2.

I suspect what happened is that in NetworkX 2.2 there was no validation of keyword arguments for the drawing functions (e.g. it looks like some explicit validation was added in #4033) so any misnamed kwarg would've been passed through without raising an exception (one of the dangers of using `**kwargs`).

**Comment by jolespin** [comment]
I should have put an actual list of edges in the example. My actual code had a list of edges and not a list of nodes.
